File: examples/governance_static_go/domain/derive_model.go
package domain

import (
	"time"
)

type DeriveModel struct {
	Name string `json:"name"`
	Version string `json:"version"`
	ParentModelId string `json:"parent_model_id"`
	DerivationType string `json:"derivation_type"`
	Description string `json:"description"`
}

func (c DeriveModel) CommandName() string { return "DeriveModel" }

func (c DeriveModel) Execute(repo AiModelRepository) (*AiModel, *DerivedModel, error) {
	agg := NewAiModel(c.Name, c.Version, "", c.Description, "", time.Time{}, c.ParentModelId, c.DerivationType, nil, nil, "")
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := DerivedModel{
		AggregateID: agg.ID,
		Name: c.Name,
		Version: c.Version,
		ParentModelId: c.ParentModelId,
		DerivationType: c.DerivationType,
		Description: c.Description,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}

File: examples/governance/model_registry_domain/lib/model_registry_domain/ai_model/commands/derive_model.rb
module ModelRegistryDomain
  class AiModel
    module Commands
      class DeriveModel
        include Hecks::Command
        emits "DerivedModel"

        attr_reader :name
        attr_reader :version
        attr_reader :parent_model_id
        attr_reader :derivation_type
        attr_reader :description

        def initialize(
          name: nil,
          version: nil,
          parent_model_id: nil,
          derivation_type: nil,
          description: nil
        )
          @name = name
          @version = version
          @parent_model_id = parent_model_id
          @derivation_type = derivation_type
          @description = description
        end

        def call
          AiModel.new(
            name: name,
            version: version,
            description: description,
            parent_model_id: parent_model_id,
            derivation_type: derivation_type,
            registered_at: Time.now.to_s,
            status: "draft"
          )
        end
      end
    end
  end
end
